fennel/testing/execute_aggregation.py

```diff
@@ -326,47 +329,240 @@ def get_val(self) -> List:
         return self.vals[min(len(self.vals) - 1, max(index - 1, 0))]
 
 
-def get_aggregated_df(
+def get_timestamps_for_hopping_window(
+    timestamp: datetime, duration: int, stride: int
+) -> List[datetime]:
+    """
+    Given a window duration, stride and a timestamp first fetch all the windows in which the
+    given timestamp lies then return the list of window end timestamps.
+    """
+    current_ts = int(timestamp.timestamp())
+
+    # Get the window of which current timestamp is part of
+    start_ts = (((current_ts - duration) // stride) + 1) * stride
+
+    results = []
+
+    # Getting all the window of which timestamp is a part of
+    while start_ts <= current_ts:
+        results.append(
+            datetime.fromtimestamp(start_ts + stride, tz=timezone.utc)
+        )
+        start_ts = start_ts + stride
+
+    return results
+
+
+def get_timestamps_for_session_window(
+    df: pd.DataFrame, key_fields: List[str], ts_field: str, gap: int
+) -> pd.DataFrame:
+    """
+    Given a session gap first the session in which the given timestamp lies then return end timestamp of the session.
+    """
+    df = df.sort_values(by=ts_field).reset_index(drop=True)
+    sessions: List[dict] = []
+    curr_session = None
+    for _, row in df.iterrows():
+        timestamp = int(row[ts_field].timestamp())
+        if curr_session is None:
+            curr_session = {"ts": timestamp, "rows": [row]}
+        else:
+            if timestamp <= curr_session["ts"] + gap:
+                curr_session["ts"] = timestamp
+                curr_session["rows"].append(row)
+            else:
+                sessions.append(curr_session)
+                curr_session = {"ts": timestamp, "rows": [row]}
+    if curr_session is not None:
+        sessions.append(curr_session)
+    rows = []
+    for session in sessions:
+        timestamp = datetime.fromtimestamp(session["ts"], tz=timezone.utc)  # type: ignore
+        for row in session["rows"]:
+            row[ts_field] = timestamp
+            rows.append(row)
+    return pd.DataFrame(rows).drop(columns=key_fields)
+
+
+def preprocess_df(
     input_df: pd.DataFrame,
     aggregate: AggregateType,
-    ts_field: str,
     key_fields: List[str],
-    output_dtype: Type,
+    ts_field: str,
 ) -> pd.DataFrame:
```

Comment:
This function impl is so long. Can we break it up ?

Reply:
broke up